TypedSyntax/README.md

```diff
@@ -0,0 +1,156 @@
+   1:1  │    f
+   1:3  │    x                                   │Float64
+   1:6  │    y                                   │Int64
+   1:9  │    z                                   │Float32
+   1:13 │  [call-i]                              │Float64
+   1:14 │    x                                   │Float64
+   1:16 │    +
+   1:17 │    [call-i]                            │Float32
+   1:18 │      y                                 │Int64
+   1:20 │      *
+   1:22 │      z                                 │Float32
+```
+
+The right hand column is the new information added by `TypedSyntaxNode`, indicating the type assigned to each variable or function call.
+
+You can also display this in a form closer to the original source code, but with type-annotations:
+
+```julia
+julia> printstyled(stdout, node; hide_type_stable=false)
+f(x::Float64, y::Int64, z::Float32)::Float64 = (x::Float64 + (y::Int64 * z::Float32)::Float32)::Float64
+```
+
+`hide_type_stable=true` (which is the default) will suppress printing of concrete types, so you need to set it to `false` if you want to see all the types.
+
+The default is aimed at identifying sources of "type instability" (poor inferrability):
+
+```julia
+julia> printstyled(stdout, TypedSyntaxNode(f, (Float64, Int, Real)))
+```
+
+which produces
+
+<code>f(x, y, z::<b>Real</b>)::<b>Any</b> = (x + (y * z::<b>Real</b>)::<b>Any</b>)::<b>Any</b></code>
+
+The boldfaced text above is typically printed in color in the REPL:
+
+- red indicates non-concrete types
+- yellow indicates a "small union" of concrete types. These usually pose no issues, unless there are too many combinations of such unions.
+
+Printing with color can be suppressed with the keyword argument `iswarn=false`.
+
+## Caveats
+
+TypedSyntax aims for accuracy, but there are a number of factors that pose challenges.
+First, anonymous and internal functions appear as part of the source text, but internally Julia handles these as separate type-inferred methods, and these are hidden from the annotator.
+Therefore, in
+
+```julia
+julia> sumfirst(c) = sum(x -> first(x), c);    # better to use `sum(first, c)` but this is just an illustration
+
+julia> printstyled(stdout, TypedSyntaxNode(sumfirst, (Vector{Any},)))
+sumfirst(c)::Any = sum(x -> first(x), c)::Any
+```
+
+`x` and `first(x)` both have type `Any`, but they are not annotated as such because they are hidden inside the anonymous function.
+
+Second, this package works by attempting to "reconstruct history": starting from the type-inferred code, it tries to map calls back to the source. It would be much safer to instead keep track of the source during inference, but at present this is not possible (see [this Julia issue](https://github.com/JuliaLang/julia/issues/31162)). There are cases where this mapping fails: for example, with
+
+```julia
+julia> function summer(list)
+           s = 0
+           for x in list
+               s += x
+           end
+           return s
+       end;
```

Comment:
It seems like we need to remove this semicolon otherwise we can't construct `TypedSyntaxNode`:
```julia
julia> function summer(list)
           s = 0
           for x in list
               s += x
           end
           return s
       end;

julia> TypedSyntaxNode(summer, (Vector{Float64},))
ERROR: MethodError: no method matching iterate(::Nothing)

Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen})
   @ Base range.jl:887
  iterate(::Union{LinRange, StepRangeLen}, ::Integer)
   @ Base range.jl:887
  iterate(::T) where T<:Union{Base.KeySet{<:Any, <:Dict}, Base.ValueIterator{<:Dict}}
   @ Base dict.jl:716
  ...

Stacktrace:
 [1] indexed_iterate(I::Nothing, i::Int64)
   @ Base ./tuple.jl:93
 [2] JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}(f::Any, t::Any; kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ TypedSyntax ~/julia/packages/Cthulhu/TypedSyntax/src/node.jl:20
 [3] JuliaSyntax.TreeNode{TypedSyntax.TypedSyntaxData}(f::Any, t::Any)
   @ TypedSyntax ~/julia/packages/Cthulhu/TypedSyntax/src/node.jl:18
 [4] top-level scope
   @ REPL[33]:1
```